daemon/containerd/image_builder.go

```diff
@@ -248,41 +250,36 @@ func (rl *rolayer) DiffID() layer.DiffID {
 }
 
 func (rl *rolayer) Release() error {
-	snapshotter := rl.c.SnapshotService(rl.snapshotter)
-	err := snapshotter.Remove(context.TODO(), rl.key)
-	if err != nil && !cerrdefs.IsNotFound(err) {
-		return err
-	}
-
-	if rl.root == "" { // nothing to release
-		return nil
-	}
-	if err := mount.UnmountAll(rl.root, 0); err != nil {
-		log.G(context.TODO()).WithError(err).WithField("root", rl.root).Error("failed to unmount ROLayer")
-		return err
-	}
-	if err := os.Remove(rl.root); err != nil {
-		log.G(context.TODO()).WithError(err).WithField("dir", rl.root).Error("failed to remove mount temp dir")
-		return err
+	if rl.lease != nil {
+		lm := rl.c.LeasesService()
+		err := lm.Delete(context.TODO(), *rl.lease)
+		if err != nil {
+			return err
+		}
+		rl.lease = nil
 	}
-	rl.root = ""
 	return nil
 }
 
 // NewRWLayer creates a new read-write layer for the builder
-func (rl *rolayer) NewRWLayer() (builder.RWLayer, error) {
+func (rl *rolayer) NewRWLayer() (_ builder.RWLayer, outErr error) {
 	snapshotter := rl.c.SnapshotService(rl.snapshotter)
 
-	// we need this here for the prepared snapshots or
-	// we'll have racy behaviour where sometimes they
-	// will get GC'd before we commit/use them
-	ctx, _, err := rl.c.WithLease(context.TODO(), leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
+	key := stringid.GenerateRandomID()
+
+	ctx, lease, err := createLease(context.TODO(), rl.c.LeasesService())
 	if err != nil {
-		return nil, fmt.Errorf("failed to create lease for commit: %w", err)
+		return nil, err
 	}
+	defer func() {
+		if outErr != nil {
+			if err := rl.c.LeasesService().Delete(ctx, lease); err != nil {
```

Comment:
Really silly question, but do you know why we need _both_ the `context` _and_ the lease, if the context has the lease(ID) attached?

Reply:
The lease is attached to the context, but is not tied to the lifetime of the context.
`Delete` deletes the specified lease and doesn't consider the lease attached to the context as a target of the operation.
The context here is just for the `Delete` operation and is not related to the lease itself.

Reply:
There's really only one way - `LeasesService.Delete`, it's just that in the usual case (`containerd.Client.WithLease`) the `Delete` is just hidden inside the cleanup function: https://github.com/containerd/containerd/blob/d015c99b2ec990c914a4b4546ec10d61cd947ab0/lease.go#L52